### PR TITLE
Refactor the cifar dataset loading code according to DRY.

### PR DESCRIPTION
**Pull request by zori**
Avoid having separate and similar retrieval functions for cifar-10 and cifar-100.

### CODE REVIEW

**Comment by t-abe** [comment]
LGTM :smile:

**Comment by unnonouno** [comment]
I can merge it after #1515 is merged.

File: chainer/datasets/cifar.py [hunk 1]
```diff
@@ -81,11 +76,62 @@ def get_cifar100(withlabel=True, ndim=3, scale=1.):
         datasets are arrays of images.
 
     """
-    raw = _retrieve_cifar_100()
+    return _get_cifar('cifar-100', withlabel, ndim, scale)
+
+
+def _get_cifar(name, *args):
```
Comment (unnonouno) [inline chainer/datasets/cifar.py]: Expand `args` to `withlabel, ndim, scale`.

File: chainer/datasets/cifar.py [hunk 2]
```diff
@@ -81,11 +76,62 @@ def get_cifar100(withlabel=True, ndim=3, scale=1.):
...
+            train_y = train_y.reshape(50000)
+        else:
+            # name == 'cifar-100'
+            def load(archive, file_name):
+                d = _pickle_load(archive.extractfile(file_name))
+                x = d['data'].reshape((-1, 3072))
+                y = numpy.array(d['fine_labels'], dtype=numpy.uint8)
+                return x, y
+
+            archive_path = download.cached_download(url)
```
Comment (unnonouno) [inline chainer/datasets/cifar.py]: This line is redundant.

File: chainer/datasets/cifar.py [hunk 3]
```diff
@@ -106,69 +152,8 @@ def _preprocess_cifar(images, labels, withlabel, ndim, scale):
...
-        train_x = train_x.reshape(50000, 3072)
-        train_y = train_y.reshape(50000)
-
-        numpy.savez_compressed(path, train_x=train_x, train_y=train_y,
-                               test_x=test_x, test_y=test_y)
-        return {'train_x': train_x, 'train_y': train_y,
-                'test_x': test_x, 'test_y': test_y}
-
-    return download.cache_or_load_file(path, creator, numpy.load)
+_root = download.get_dataset_directory(os.path.join('pfnet', 'chainer',
```
Comment (unnonouno) [inline chainer/datasets/cifar.py]: Do we need to call this method here? I feel it is enough to call it in `_get_cifar`.

**Comment by okuta** [comment]
@zori Do you have time for this PR?

File: chainer/datasets/cifar.py (hunk 2 above)
Comment (zori) [inline chainer/datasets/cifar.py]: Included in the following `with` statement (I had added it for clarity).

File: chainer/datasets/cifar.py (hunk 1 above)
Comment (zori) [inline chainer/datasets/cifar.py]: Done.

File: chainer/datasets/cifar.py (hunk 3 above)
Comment (zori) [inline chainer/datasets/cifar.py]: Moved in `_get_cifar`.

**Comment by zori** [comment]
@okuta I fixed what Unno-san recommended above. nosetests from `tests/chainer_tests/datasets_tests/test_cifar.py` pass. I apologize for being late with this PR.

File: chainer/datasets/cifar.py (hunk 2 above)
Comment (unnonouno) [inline chainer/datasets/cifar.py]: Ah, sorry. I mean, `archive_path` is assigned at line 87.

Comment (unnonouno) [inline chainer/datasets/cifar.py]: Ah, sorry. I mean, archive_path is assigned at line 87.

File: chainer/datasets/cifar.py [hunk 4]
```diff
@@ -81,11 +76,64 @@ def get_cifar100(withlabel=True, ndim=3, scale=1.):
...
+            train_y = train_y.reshape(50000)
+        else:
+            # name == 'cifar-100'
+            def load(archive, file_name):
+                d = _pickle_load(archive.extractfile(file_name))
+                x = d['data'].reshape((-1, 3072))
+                y = numpy.array(d['fine_labels'], dtype=numpy.uint8)
+                return x, y
+
+            with tarfile.open(download.cached_download(url), 'r:gz') as \
```
Comment (unnonouno) [inline chainer/datasets/cifar.py]: `url` is already downloaded at line 89, isn't it?

Comment (zori) [inline chainer/datasets/cifar.py]: I see, obsolete, indeed. I fixed it.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**